kstatus/README.md

```diff
@@ -0,0 +1,116 @@
+builds upon the basics to provide a higher level API.
+
+**sigs.k8s.io/kustomize/kstatus/status**: Provides two basic functions. First, it provides the `Compute` function
+that takes a single resource and computes the status for this resource based on the fields in the status object for
+the resource. Second, it provides the `Augment` function that computes the appropriate standard conditions based on
+the status object and then amends them to the conditions in the resource. Both of these functions currently operate
+on Unstructured types, but this should eventually be changed to rely on the kyaml library. Both of these functions 
+compute the status and conditions solely based on the data in the resource passed in. It does not communicate with
+a cluster to get the latest state of the resources.
+
+**sigs.k8s.io/kustomize/kstatus/wait**: This package builds upon the status package and provides functionality that
+will fetch the latest state from a cluster. It provides the `FetchAndResolve` function that takes list of resource
+identifiers, fetches the latest state for all the resources from the cluster, computes the status for all of them and
+returns the results. `WaitForStatus` accepts a list of resource identifiers and will poll cluster for the status of 
+the resources until all resources have reached the `Current` status.
+
+## Challenges
+
+### Status is not obvious for all resource types
+
+For some types of resources, it is pretty clear what the different statuses mean. For others, it
+is far less obvious. For example, what does it mean that a PodDisruptionBudget is Current? Based on
+the assumptions above it probably should be whenever the controller has observed the resource
+and updated the status object of the PDB with information on allowed disruptions. But currently, a PDB is
+considered Current when the number of healthy replicas meets the threshold given in the PDB. Also, should
+the presence of a PDB influence when a Deployment is considered Current? This would mean that a Deployment
+should be considered Current whenever the number of replicas reach the threshold set by the corresponding
+PDB. This is not currently supported as described below.
+
+### Status is decided based on single resource
+Currently the status of a resource is decided solely based on information from
+the state of that resource. This is an issue for resources that create other resources
+and that doesn't provide sufficient information within their own status object. An example
+is the Service resource that doesn't provide much status information but do generate Endpoint
+resources that could be used to determine status. Similar, the status of a Deployment could be 
+based on its generated ReplicaSets and Pods.
+
+Not having the generated resources also limits the amount of details that can be provided
+when something isn't working as expected.
```

Comment:
_Settled_ (controller loop reconciled) and _Healthy_ (server responding to requests from outside world / probes happy) are two orthogonal independent variables, each with a different domain of possible values.

Are we conflating them?  Seems like we are, but I don't have a crisp understanding of it.  Imagine you're debugging a production problem.  Wouldn't you want to separate indicators for these things, so you'd know where to start investigating?

Reply:
So I agree these are two different aspects and for this effort we are only considering whether the controller loop has reconciled, i.e. whether everything looks good from the perspective of the control plane. If for some reason the control plane is not able to reconcile, we want the errors reported to be as actionable as possible, and without knowing the generated resources this can be difficult. For example, if a deployment can't create enough replicas due to insufficient capacity, by looking at the deployment resource we can only see that we failed to reach the desired number of replicas. If we could inspect the state of the replicasets and pods, we could actually understand why and report that the cause is insufficient cpu/memory and suggest scaling up the cluster.

Reply:
sg, thanks